src/macro.jl

```diff
@@ -0,0 +1,116 @@
+# We define a custom subtype of `AbstractIrrational` and
+# define methods for it that are not generalized yet to `AbstractIrrational`
+# in https://github.com/JuliaLang/julia/blob/e536c77f4dc693aafc48af910b4fd86b487e900d/base/irrationals.jl
+
+abstract type IrrationalConstant <: AbstractIrrational end
+
+function Base.show(io::IO, ::MIME"text/plain", x::IrrationalConstant)
+    if get(io, :compact, false)::Bool
+        print(io, x)
+    else
+        print(io, x, " = ", string(float(x))[1:min(end,15)], "...")
+    end
+end
+
+Base.:(==)(::T, ::T) where {T<:IrrationalConstant} = true
+Base.:<(::T, ::T) where {T<:IrrationalConstant} = false
+Base.:<=(::T, ::T) where {T<:IrrationalConstant} = true
+Base.hash(x::IrrationalConstant, h::UInt) = 3*objectid(x) - h
+
+# definitions for AbstractIrrational added in https://github.com/JuliaLang/julia/pull/34773
+if VERSION < v"1.5.0-DEV.301"
+    Base.zero(::IrrationalConstant) = false
+    Base.zero(::Type{<:IrrationalConstant}) = false
+
+    Base.one(::IrrationalConstant) = true
+    Base.one(::Type{<:IrrationalConstant}) = true
+end
+
+# definition for AbstractIrrational added in https://github.com/JuliaLang/julia/pull/31068
+if VERSION < v"1.2.0-DEV.337"
+    Base.inv(x::IrrationalConstant) = 1/x
+end
+
+"""
+    @irrational sym val def [T]
+    @irrational(sym, val, def[, T])
+
+Define a new singleton type `T` representing an irrational constant as subtype of
+`AbstractIrrational` with an instance named `sym`, pre-computed `Float64` value `val`,
```

Comment:
```suggestion
`IrrationalConstant` with an instance named `sym`, pre-computed `Float64` value `val`,
```

Reply:
I think we might want to remove `IrrationalConstant` when these methods are generalized to `AbstractIrrational` in base but I'm not entirely sure. I don't have a strong opinion on the text but to me `IrrationalConstant` seemed more like an internal implementation detail, and in contrast to `AbstractIrrational` it is not exported.

Reply:
I see.
I feel `IrrationalConstant` is not just an internal implementation because some other packages might add irrational constants to this abstract type. This is just a docstring, so explicit `IrrationalConstant` seems better for me.

Reply:
Let's be completely clear:
```suggestion
`IrrationalConstants.IrrationalConstant <: AbstractIrrational` with an instance named `sym`, pre-computed `Float64` value `val`,
```

Reply:
Do you think the docstring is sufficiently clear this way @hyrodium?